src/boot/CompressedAsyncStorage.js

```diff
@@ -0,0 +1,46 @@
+/* @flow */
+import { AsyncStorage, NativeModules } from 'react-native';
+
+export default class CompressedAsyncStorage {
+  // HACK: This wrapper around AsyncStorage.getItem() is not perfectly compliant with the
+  // specification of AsyncStorage.getItem(): `callback` should be optional, and getItem()
+  // should return the result. Here, we declare `callback` as non-optional, because redux-persist
+  // always calls getItem() with a callback. redux-persist is the only library that uses
+  // CompressedAsyncStorage. Also, we don't return the result, because redux-persist never uses it
+  // (instead, it looks at the result passed to `callback`). This avoids decompressing the result
+  // twice.
+  static async getItem(key: string, callback: (error: ?Error, result: ?string) => void) {
+    async function wrappedCallback(error, result) {
+      // It's possible that getItem() is called on uncompressed state (e.g. when a user updates their
+      // app from a version without compression to a version with compression). Detect if the stored
+      // state is compressed, and return it unmodified if it isn't. Detect a compressed state by
+      // inspecting the first byte of the string. TextCompressionModule uses zlib compression and returns
+      // compressed strings that start with two bytes that specify the compression method. The
+      // bytes then get base64-encoded and turn into a string starting with 'e'. Luckily, uncompressed
+      // state must be valid JSON and hence cannot start with an e. The only possible JSON start characters
+      // are:
+      // { - Object
+      // [ - Array
+      // 0123456789 - number
+      // " - string
+      // f - false
+      // t - true
+      // n - null
+      if (!result.startsWith('e')) {
+        callback(error, result);
+        return;
+      }
+      callback(error, await NativeModules.TextCompressionModule.decompress(result));
+    }
+    await AsyncStorage.getItem(key, wrappedCallback);
+  }
+  static async setItem(key: string, value: string, callback: ?(error: ?Error) => void) {
```

Comment:
Style nit: put a blank line between each of these methods. That makes it easier to visually scan.